### PR TITLE
Add Distill Bi-LSTM

### PR DESCRIPTION
**Pull request by LiuChiachi**
Add Distill Lstm

### CODE REVIEW

File: PaddleNLP/examples/distill/distill_lstm/bert_distill.py [hunk 1]
```diff
@@ -0,0 +1,267 @@
...
+        if task_name == 'qqp' or task_name == 'mnli':
+            _, _, small_input_ids_1, seq_len_1, small_input_ids_2, seq_len_2, labels = batch
+            logits = model(small_input_ids_1, seq_len_1, small_input_ids_2,
+                           seq_len_2)
+        else:
+            _, _, small_input_ids, seq_len, labels = batch
+            logits = model(small_input_ids, seq_len)
+
+        loss = alpha * ce_loss(logits, labels) + (1 - alpha) * mse_loss(
+            logits, teacher_logits)
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/bert_distill.py]: evaluate感觉只需要评估当前训练的模型就可以，应该无需考虑与teacher_logits的mse_loss

File: PaddleNLP/examples/distill/distill_lstm/bert_distill.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+                loss.numpy(),
+                res[0],
+                res[1],
+                res[2],
+                res[3],
+                res[4], ),
+            end='')
+    elif isinstance(metric, Mcc):
+        print("eval loss: %f, mcc: %s, " % (loss.numpy(), res[0]), end='')
+    elif isinstance(metric, PearsonAndSpearman):
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/bert_distill.py]: 感觉这里的metric打印略麻烦，可否根据metric中name等信息来打印，然后这里可以更简单些

File: PaddleNLP/examples/distill/distill_lstm/bert_distill.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+            % (loss.numpy(), res[0], res[1], res[2]),
+            end='')
+    else:
+        print("eval loss: %f, acc: %s, " % (loss.numpy(), res), end='')
+    model.train()
+
+
+def do_train(
+        task_name='sst-2',
+        language='en',
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/bert_distill.py]: language可以去掉，根据task_name来判断中英文任务

File: PaddleNLP/examples/distill/distill_lstm/bert_distill.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        padding_idx=0,
+        dropout_prob=0.5,
+        save_steps=20,
+        alpha=0.0,
+        model_name='bert-base-uncased',
+        teacher_path='model/SST-2/best_model_610/model_state.pdparams',
+        vocab_path='/root/.paddlenlp/models/bert-base-uncased/bert-base-uncased-vocab.txt',
+        use_pretrained_w2v=True,
+        data_augmentation=False,
+        n_iter=20):
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/bert_distill.py]: 参数都尽量使用args封装提供吧

File: PaddleNLP/examples/distill/distill_lstm/bert_distill.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+
+    metric_class = TASK_CLASSES[task_name][1]
+    metric = metric_class()
+
+    teacher = Teacher(model_name=model_name, param_path=teacher_path)
+    teacher_eval_logits_list = []
+    with paddle.no_grad():
+        for i, batch in enumerate(dev_data_loader):
+            input_ids, segment_ids = batch[:2]
+            teacher_logits = teacher.model(input_ids, segment_ids)
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/bert_distill.py]: 同上，evaluate应该主要是为了评估训练的模型在dev数据上的效果，应该不需要考虑teacher

File: PaddleNLP/examples/distill/distill_lstm/bert_distill.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+    # task_name = 'sst-2'
+    # task_name = 'qqp'
+
+    vocab_size_dict = {
+        "senta": 29496,
+        # "bert-base-chinese": 21128,
+        # "bert-base-uncased": 30522,
+        "sst-2": 30522,
+        "qqp": 30522,
+    }
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/bert_distill.py]: 代码可以再整理下，按照task提供配置

File: PaddleNLP/examples/distill/distill_lstm/run_bert_finetune.py [hunk 7]
```diff
@@ -0,0 +1,459 @@
+# Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/run_bert_finetune.py]: 是否必须再单独提个一个bert finetune的脚本呢，可以的话直接使用glue或者bert中的代码吧，README中说明使用方式即可

File: PaddleNLP/examples/distill/distill_lstm/small.py [hunk 8]
```diff
@@ -0,0 +1,269 @@
...
+            output_dim,
+            weight_attr=paddle.ParamAttr(initializer=I.Uniform(
+                low=-init_scale, high=init_scale)))
+        self.dropout = nn.Dropout(dropout_prob)
+
+    def forward(self, x_1, seq_len_1, x_2=None, seq_len_2=None):
+        x_embed_1 = self.embedder(x_1)
+        lstm_out_1, (hidden_1, _) = self.lstm(
+            x_embed_1, sequence_length=seq_len_1)
+        self.x_embed_1 = x_embed_1
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/small.py]: 注意去掉中间debug代码与注释掉的代码

File: PaddleNLP/examples/distill/distill_lstm/small.py [hunk 9]
```diff
@@ -0,0 +1,269 @@
...
+                (hidden_2[-2, :, :], hidden_2[-1, :, :]), axis=1)
+            out = paddle.concat(
+                x=[out_1, out_2, out_1 + out_2, paddle.abs(out_1 - out_2)],
+                axis=1)
+            out = paddle.tanh(self.fc_1(out))
+        else:
+            out = paddle.tanh(self.fc(out_1))
+
+        logits = self.output_layer(out)
+        logits = self.dropout(logits)
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/small.py]: 这里再确认下，对logits做dropout应该不太正确

File: PaddleNLP/examples/distill/distill_lstm/bert_distill.py (hunk 4 above)
Comment (LiuChiachi) [reply PaddleNLP/examples/distill/distill_lstm/bert_distill.py]: Thanks. Done.

File: PaddleNLP/examples/distill/distill_lstm/bert_distill.py (hunk 5 above)
Comment (LiuChiachi) [reply PaddleNLP/examples/distill/distill_lstm/bert_distill.py]: Thanks. Done.

File: PaddleNLP/examples/distill/distill_lstm/bert_distill.py (hunk 6 above)
Comment (LiuChiachi) [reply PaddleNLP/examples/distill/distill_lstm/bert_distill.py]: Thanks. Done.

File: PaddleNLP/examples/distill/distill_lstm/run_bert_finetune.py (hunk 7 above)
Comment (LiuChiachi) [reply PaddleNLP/examples/distill/distill_lstm/run_bert_finetune.py]: 感谢建议，这里不需要bert finetune的完整脚本了，已经把修改过的`convert_example` move到`utils.py`里

File: PaddleNLP/examples/distill/distill_lstm/small.py (hunk 8 above)
Comment (LiuChiachi) [reply PaddleNLP/examples/distill/distill_lstm/small.py]: Done. Thanks

File: PaddleNLP/examples/distill/distill_lstm/small.py (hunk 9 above)
Comment (LiuChiachi) [reply PaddleNLP/examples/distill/distill_lstm/small.py]: 感谢，已经改正

File: PaddleNLP/examples/distill/distill_lstm/bert_distill.py [hunk 10]
```diff
@@ -0,0 +1,153 @@
...
+from data import create_distill_loader, load_embedding
+
+TASK_CLASSES = {
+    "sst-2": (GlueSST2, Accuracy),
+    "qqp": (GlueQQP, AccuracyAndF1),
+    "senta": (ChnSentiCorp, Accuracy),
+}
+
+
+class Teacher(object):
```
Comment (ZeyuChen) [inline PaddleNLP/examples/distill/distill_lstm/bert_distill.py]: TeacherModel可能更为精准。

File: PaddleNLP/examples/distill/distill_lstm/README.md [hunk 11]
```diff
@@ -0,0 +1,142 @@
+# Distill Bi-LSTM
```
Comment (ZeyuChen) [inline PaddleNLP/examples/distill/distill_lstm/README.md]: Distil-BiLSTM是否更为合适？我记得大部分命名规则都似乎DistilBERT Distil-LSTM，似乎没有双写l, 这块可以再确认下

File: PaddleNLP/examples/distill/distill_lstm/small.py [hunk 12]
```diff
@@ -0,0 +1,203 @@
+# Copyright (c) 2021 PaddlePaddle Authors. All Rights Reserve.
```
Comment (ZeyuChen) [inline PaddleNLP/examples/distill/distill_lstm/small.py:1]: small这个文件名是否清晰呢？student.py是否会更好？

**Review by ZeyuChen — COMMENTED** [review]
一些小意见。

File: PaddleNLP/examples/distill/distill_lstm/data.py [hunk 13]
```diff
@@ -0,0 +1,367 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import io
+import os
+from functools import partial
+import gensim
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/data.py:19]: README中说明依赖

File: PaddleNLP/examples/distill/distill_lstm/README.md [hunk 14]
```diff
@@ -0,0 +1,142 @@
...
+```
+.
+├── small.py              # 小模型结构以及对小模型的训练脚本
+├── bert_distill.py       # 用大模型bert蒸馏小模型的蒸馏脚本
+├── data.py               # 定义了dataloader等数据读取接口
+├── utils.py              # 定义了将样本转成id的转换接口
+├── args.py               # 参数配置脚本
+└── README.md             # 文档，本文件
+```
+
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/README.md:14]: README有点过于简单了，参考了什么论文、使用了什么方法、和论文中数据增强部分的差异都没有给出

File: PaddleNLP/examples/distill/distill_lstm/README.md [hunk 15]
```diff
@@ -0,0 +1,142 @@
...
+    --batch_size 128   \
+    --learning_rate 3e-5 \
+    --num_train_epochs 3 \
+    --logging_steps 10 \
+    --save_steps 10 \
+    --output_dir ../distill/ditill_lstm/model/$TASK_NAME/ \
+    --n_gpu 1 \
+
+```
+训练完成之后，可将训练效果最好的模型保存在该项目下的`models/$TASK_NAME/`下。模型目录下有`model_config.json`, `model_state.pdparams`, `tokenizer_config.json`及`vocab.txt`这几个文件。
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/README.md]: 更多说明详见example/glue，给出连接。

File: PaddleNLP/examples/distill/distill_lstm/README.md [hunk 16]
```diff
@@ -0,0 +1,142 @@
...
+
+### 数据准备
+
+在中文数据集上的小模型训练的输入是利用jieba分词的，词表是由senta提供，可通过运行以下命令进行下载
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/README.md]: `词表是由senta提供`  中文数据同example/senta，注意给出链接。

另外注意顺序，先是数据准备，然后是训练

File: PaddleNLP/examples/distill/distill_lstm/README.md [hunk 17]
```diff
@@ -0,0 +1,142 @@
...
+    --task_name sst-2 \
+    --dropout_prob 0.2 \
+    --batch_size 128 \
+    --model_name bert-base-uncased \
+    --use_pretrained_emb True \
+    --teacher_path model/SST-2/best_model_610/model_state.pdparams
+```
+
+```shell
+CUDA_VISIBLE_DEVICES="5" python bert_distill.py \
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/README.md]: 感觉给出CUDA_VISIBLE_DEVICES="5"这些到5可能不太好

File: PaddleNLP/examples/distill/distill_lstm/bert_distill.py (hunk 10 above)
Comment (LiuChiachi) [reply PaddleNLP/examples/distill/distill_lstm/bert_distill.py]: 感谢，已经修改

File: PaddleNLP/examples/distill/distill_lstm/README.md (hunk 11 above)
Comment (LiuChiachi) [reply PaddleNLP/examples/distill/distill_lstm/README.md]: 这个论文作者没有给这个蒸馏模型起名字，我现在把这个`README`的大标题修改成了` Distilling knowledge from fine-tuned BERT into Bi-LSTM`，是否合适呢

File: PaddleNLP/examples/distill/distill_lstm/small.py (hunk 12 above)
Comment (LiuChiachi) [reply PaddleNLP/examples/distill/distill_lstm/small.py:1]: 我理解student的概念应该和teacher成对出现，目前叫`small.py`的文件只用于单独训`bilstm`，只为和蒸馏后的效果作对比，没有体现师生的感觉。模型的蒸馏提现在`bert_distill.py`中，在这个文件里我把small都替换成了student。这样是否恰当呢，感谢指点~

File: PaddleNLP/examples/distill/distill_lstm/data.py (hunk 13 above)
Comment (LiuChiachi) [reply PaddleNLP/examples/distill/distill_lstm/data.py:19]: 在README进行了补充说明，感谢指出

File: PaddleNLP/examples/distill/distill_lstm/README.md (hunk 14 above)
Comment (LiuChiachi) [reply PaddleNLP/examples/distill/distill_lstm/README.md:14]: 感谢，已经对README进行了补充。

File: PaddleNLP/examples/distill/distill_lstm/README.md (hunk 17 above)
Comment (LiuChiachi) [reply PaddleNLP/examples/distill/distill_lstm/README.md]: 感谢，已经修改

File: PaddleNLP/examples/distill/distill_lstm/README.md (hunk 16 above)
Comment (LiuChiachi) [reply PaddleNLP/examples/distill/distill_lstm/README.md]: 感谢。已经对README进行了修改

File: PaddleNLP/examples/distill/distill_lstm/data.py [hunk 18]
```diff
@@ -0,0 +1,367 @@
...
+
+    for idx in range(len(tokenized_text)):
+        new_data.append(tokenized_text[idx])
+        for _ in range(n_iter):
+            # 1. Masking
+            words = [
+                tokenizer.unk_token if np.random.rand() < p_mask and
+                not word.startswith('##') else word
+                for word in tokenized_text[idx][0]
+            ]
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/data.py:106]: Mask这块最好还是完善下，感觉还是在tokenize前选择要MASK的word，在tokenize之后去MASK掉该word包含的所有subword。

File: PaddleNLP/examples/distill/distill_lstm/data.py [hunk 19]
```diff
@@ -0,0 +1,367 @@
...
+    new_data = []
+    for data in train_dataset:
+        new_data.append(data)
+
+        for _ in range(n_iter):
+            # 1. Masking
+            words = [
+                "MASK" if np.random.rand() < p_mask else word
+                for word in jieba.cut(data[0])
+            ]
```
Comment (guoshengCS) [inline PaddleNLP/examples/distill/distill_lstm/data.py:164]: 中文这里也最好完善下，虽然起到的是按照词来mask的作用，但是利用jieba将MASK产生[UNK]，teacher模型那边利用BERTTokenzier的[UNK]这种方式也不是特别合适。

**Review by guoshengCS — APPROVED** [review]
